### DISCUSSION TITLE
OAD Structure and referencing in OpenAPI 3.1

### DISCUSSION THREAD

**Discussion by mikekistler** [OAI/OpenAPI-Specification#4114] [General]
We've been having a lot of discussions about document structure and how OAS and JSON Schema references work in various places, such as

- In recent TDC meetings
- In PR #4100, which is for 3.0 but winds up discussing how 3.1 works in places
- In Slack, particularly [this _looooong_ thread in the spec channel](https://open-api.slack.com/archives/C1137F8HF/p1726772743170639)

I think we can continue discussing the 3.0 behavior in #4100, but (to the degree that these can be separated) thought it would be a good idea to start a discussion focused primarily on the behavior in 3.1, which _I think_ most folks feel is what we want going forward.

**Comment by mikekistler** [comment]
I'll try to summarize my understanding of how document structure and referencing works in OAS 3.1, quoting the 3.1.1 spec or borrowing words from @handrews and others in places.

From the [3.1.0 spec](https://github.com/OAI/OpenAPI-Specification/blob/3.1.0/versions/3.1.0.md#document-structure), the general concept is this:
> An OpenAPI document MAY be made up of a single document or be divided into multiple, connected parts at the discretion of the author.

As for how references work, the [3.1.0 spec](https://github.com/OAI/OpenAPI-Specification/blob/3.1.0/versions/3.1.0.md#relative-references-in-uris) says this:
> Unless specified otherwise, all properties that are URIs MAY be relative references as defined by [RFC3986](https://tools.ietf.org/html/rfc3986#section-4.2).

which was also the case in OAS 3.0, but the difference in 3.1 is that URI references are resolved differently. In OAS 3.1:
> Relative references [in OAS fields --not in Schema Objects] are resolved using the referring document as the Base URI according to [RFC3986](https://tools.ietf.org/html/rfc3986#section-5.2).

and

> Relative references in Schema Objects including any that appear as $id values, use the nearest parent $id as a Base URI, as described by [JSON Schema Specification Draft 2020-12](https://tools.ietf.org/html/draft-bhutton-json-schema-00#section-8.2).

There's more to the story but I'm going to save this and amend with more content later.

**Reply by handrews** [reply]
In 3.0 they are URL-references rather than URI-references in 3.1.  That's why "Resolving Relative Refrerences in URLss" was split into two sections.

BTW, several things in 3.1.0  are misleading or (like the "identifies the location" text spliced in from JSON Reference when we removed JSON Reference) outright wrong (that phras contradicted the implications of "Resolving Relative References in URIs", so one of them had to be wrong and we decided it was that phrase).  So I don't think reading 3.1.0 is going to give you the right impression throughout.  The current 3.1.1 text is what you want to start with.

**Reply by mikekistler** [reply]
Before tackling references from Schema Objects, I think we should explore how references outside of Schema Objects work. I'm a little surprised by the wording on this, which appears to be preserved in [OAS 3.1.1](https://github.com/OAI/OpenAPI-Specification/blob/v3.1.1-dev/versions/3.1.1.md#relative-references-in-api-description-uris) though with significant elaboration. Again:
> Relative references [in OAS fields --not in Schema Objects] are resolved using the referring document as the Base URI 

As a concrete example, let's consider a Reference Object that may appear as an element in the array value of the [`parameters` field of an Operation Object](https://github.com/OAI/OpenAPI-Specification/blob/3.1.0/versions/3.1.0.md#fixed-fields-8).

If this reference is only a fragment, then no problem -- the base URI is the "referring document" which is the same as the "referenced document".

But now suppose that the $ref is something like "otherDoc.json#/components/parameters/myParam.json".  And suppose the "components/parameters/myParam.json" fragment in OtherDoc.json contains this parameter definition:

```
    {
        "name":  "myParam",
        "in": "query",
        "schema": {
            "$ref": "#/components/schemas/ParamSchema"
        }
    }
```

Since the Base URI for this reference is the _referring_ document, I think that means that we must use the "ParamsSchema" in the components/schemas of the _referring_ document -- and not the referenced document -- when resolving this reference.

Have I understood this correctly?

**Reply by handrews** [reply]
@mikekistler The "referring document" is _always_ the document containing the `$ref` (or more accurately, containing the URI-reference).  Each relative URI-reference is evaluated on its own.  It does not matter how you reached it.  (In the Schema Object, it matters how you reach `$dynamicRef`, which is what makes it dynamic).

**Reply by handrews** [reply]
Also, sometimes refs in Schema Objects use the document's retrieval URI.

One point of confusion here is that the OAS, in all versions, tries to over-specify the base URI calculation which, aside from whether §5.1.1 applies, is entirely governed by RFC3986 and doesn't really need further elaboration.  JSON Schema's `$id` (and the proposed `self` field for 3.2) are §5.1.1 mechanisms, and then everything else just flows from RFC3986 as written.  It would be better if we said less, probably, but I was trying not to change too much.

But I think this is making you think it's more complex than it is.  All that happens here is that in Schema Objects, there's a 5.1.1 source, and there's encapsulation relevant to 5.1.2.  That's it.  Nothing else is under JSON Schema or the OAS's control.

**Reply by mikekistler** [reply]
Thanks @handrews -- I _think_ you have confirmed that my understanding is correct. Which means that in OAS 3.1 $refs outside of Schema Objects work _completely different_ from $refs inside Schema Objects. I'm surprised by this but this whole topic has been a huge education for me.

**Reply by handrews** [reply]
@mikekistler 

> Which means that in OAS 3.1 $refs outside of Schema Objects work completely different from $refs inside Schema Objects

Hmm... that's not really how I think about it.  We have 8 fields that use URI-references.  Six of them are what I think of as "normal" URI-reference fields:

* Reference Object `$ref`
* Path Item Object `$ref`
* Schema Object `$ref`
* Link Object `operationRef`
* Example Object `externalValue`
* External Documentation Object `url`

There are then two fields that require a preprocessing step:
* Discriminator Object `mapping`, which has to first figure out whether the value is a component name of a URI-reference, but if it's a URI-reference it is otherwise normal
* `$dynamicRef`, which we've discussed elsewhere and requires a pre-processing step to determine the resource that provides the base URI — this is weird and I regret it and it has _no impact_ on how anything else behaves at all, so I'm going to ignore it for the rest of this comment.

So let's think of our seven normal URI-reference fields (the six totally normal ones, and the URI-reference syntax of `mapping`).  There are three aspects of how these function:

1.  Determine the base URI, which is _always_ done in accordance with RFC3986, which allows customizing the first possible base URI source (per §5.1.1) based on the resource's media type, after which §5.1.2 – 5.1.4 are the same for all media types
2. Resolve the URI-reference against the base URI to produce a full URI (if the URI-reference is not relative, this is a no-op, but technically still occurs, and again every aspect of this is defined by RFC3986 and we do not change it- none of our fields _other_ than `$dynamicRef` even attempt to change this, with preprocessing steps or otherwise)
3. Handling the reference target:  This is where these fields differ, in terms of how adjacent fields are handled and whether there is an inline equivalent or whether the effect is purely delegation.  This is the _only_ aspect in which the fields differ (assuming we've already determined `mapping` is using the URI-reference syntax)

Where you seem to be seeing a difference is in the base URI determination, but these are just determined by the (unregistered) media types (with the YAML equivalents behaving the same as the JSON ones):

* `application/openapi+json`, as of OAS 3.1, does not define a §5.1.1 mechanism for base URI determination, leaving it entirely determined by §5.1.2-5.1.4 (as of OAS 3.2, it will define `self` in the OpenAPI Object as its §5.1.1 mechanism)
* `application/schema+json` uses `$id` in a [schema resource root](https://www.ietf.org/archive/id/draft-bhutton-json-schema-01.html#name-root-schema-and-subschemas-) as its §5.1.1 mechanism, _and_ allows embedding multiple schema resources in a single document (every schema document is a schema resource, but not all schema resources are separate schema documents)

But none of that changes how any `$ref` or any of these other "normal" URI-reference keywords function.  They all just resolve against the base URI, whatever it is.  Figuring out the base URI and resolving a URI-reference against it are separate steps: the first is impacted by the media type under §5.1.1, and the second is absolutely standard in all cases for these seven fields.

**Reply by mikekistler** [reply]
Hmmm ... now I think you are _disagreeing_ with my earlier statement. Maybe I need a better / more complete example. I'll work on that.

**Reply by mikekistler** [reply]
Okay ... maybe I don't need a more complete example, as I think I see where my confusion is.

I think the key is understanding what "the referring document" is this statement:
> Relative references [in OAS fields --not in Schema Objects] are resolved using the referring document as the Base URI according to [RFC3986](https://tools.ietf.org/html/rfc3986#section-5.2).

"document" is unfortunately a very generic term and could mean many things.

It seems reasonable to interpret "document" to be "JSON document". And the JSON Schema 2020-12 spec does [define "JSON document"](https://json-schema.org/draft/2020-12/draft-bhutton-json-schema-01#name-json-document) and in particular says this:
> In JSON Schema, the terms "JSON document", "JSON text", and "JSON value" are interchangeable because of the data model it defines.

So if the entry document referenced "otherDoc.json#/components/parameters/myParam.json", it seems reasonable to say that the value of that fragment, which is a JSON value, _is_ "the referring document" for any "$ref" within it.

I acknowledge that there are other, maybe better interpretations, but I'd like to know if there is any reason that this interpretation should be considered invalid.

**Reply by handrews** [reply]
@mikekistler 

> It seems reasonable to interpret "document" to be "JSON document". And the JSON Schema 2020-12 spec does [define "JSON document"](https://json-schema.org/draft/2020-12/draft-bhutton-json-schema-01#name-json-document) and in particular says this:
> 
> > In JSON Schema, the terms "JSON document", "JSON text", and "JSON value" are interchangeable because of the data model it defines.

This is about the data model, and has nothing to do with [resources vs documents](https://www.ietf.org/archive/id/draft-bhutton-json-schema-01.html#name-root-schema-and-subschemas-).

> So if the entry document referenced "otherDoc.json#/components/parameters/myParam.json", it seems reasonable to say that the value of that fragment, which is a JSON value, is "the referring document" for any "$ref" within it.

A reference connects a referring document to a referenced document.  The "referring document" is just the document containing the reference.

I don't know who added that term to 3.1 because Ron squashed a ton of commits when releasing the spec and obliterated a lot of the history, so I can't track down the historical context.  But this is not some strange new concept.  Document A refers to document B (using a `$ref`), therefore document A is the referring document.

@mikekistler A reference connects _from_ a referring document/resource _to_ a referred or reference document/resource (really, it should be resource, but in most cases they are the same, just not in JSON Schema... or in the `multipart/related` document structure with `Content-Location` that I mentioned earlier... or if the "document" is a YAML stream...)

**Comment by miqui** [comment]
tdc notes: resource , documents, and description.    (rework, should be addressed before patch release?)

**Comment by mikekistler** [comment]
We had a lively discussion of this topic in the TDC meeting yesterday.

Trying to tease apart what is meant by "OpenAPI Description" vs "OpenAPI Document" has been challenging.

And connecting these to related terms used in the OAS or referenced RFCs, such as "resource", "entity",  adds yet more complexity.

There were differing views on what should be considered "the referring document" in this sentence taken from the [OAS 3.1.0](https://github.com/OAI/OpenAPI-Specification/blob/3.1.0/versions/3.1.0.md#relative-references-in-uris):

> Relative references [not in Schema Objects] are resolved using the referring document as the Base URI according to RFC3986.

There was the view I presented in the [comment above](https://github.com/OAI/OpenAPI-Specification/discussions/4114#discussioncomment-10762938), which is basically that the JSON value found at the fragment could be considered the referring document (ignoring the outer "document" in which it was contained).

Henry argued that this interpretation is invalid -- that the referring document _must_ be considered to the document/resource referenced by the "absolute" part of the URI (fragment removed). As support for this view, Henry pointed out that fragment processing must be done based on the MIME type of the resource and "application/json" has no defined fragment identification syntax, citing RFC 6901 (section 5). As such, referencing a fragment in a "plain JSON" document is undefined and (as spec writers) we should disallow it.

At the end of the meeting we agreed to attempt to formalize the "description", "document", "resource" terminology and then use that as a way of clarifying this whole area.

**Reply by handrews** [reply]
Something I did not entirely realize until near the end of the meeting, and tried to address but I guess I was not clear, I'd that you seem to have switched "referring" and "referenced".  

The document containing the '$ref' is the referring document. No matter what definition of document is used. 

The document to which the value of the '$ref' refers is the referenced document. Again, regardless of the definition of document. 

Can we get clear agreement on that before proceeding to other aspects?

**Reply by handrews** [reply]
@mikekistler meant to tag you since discussion notifications seem a little different.

**Reply by mikekistler** [reply]
I agree that your descriptions are correct. Where I did get confused was thinking that "using the referring document as the Base URI" meant that this was the Base URI to be used _in the referenced document_. But now I realize that this means only for dereferencing the URI in the referring document. Once that URI is dereferenced, it becomes the base URI in the referenced document.

**Reply by handrews** [reply]
@mikekistler 

> But now I realize that this means only for dereferencing the URI in the referring document.

Technically, it's for _resolving_ the (possibly relative) URI-reference value to a full URI.  Then the full URI is _dereferenced_ (which might involve an intermediate step to map the identifier (URI) to a location (URL)).  (This usage of "resolving" and "dereferencing" is from RFC3986- "mapping" URIs to URLs is my own terminology to avoid overloading "resolving" to also mean that step)

> Once that URI is dereferenced, it becomes the base URI in the referenced document.

Sometimes.  If the URI is the retrieval URL of the referenced document _and_ there is no base URI for that document available under RFC3986 §5.1.1 or §5.1.2, then §5.1.3 applies and the retrieval URL is the base URI for the document (it would also be the base URI used to resolve a relative URI-reference that provides a base URI under 5.1.2, or a relative URI-reference under 5.1.1 if there is no 5.1.2 value).  This is one reason why the `self` proposal for 3.2 is important:  Without it, you can jump back and forth between identifiers and locators rather unexpectedly.

**Reply by handrews** [reply]
@mikekistler 

> At the end of the meeting we agreed to attempt to formalize the "description", "document", "resource" terminology and then use that as a way of clarifying this whole area.

* A document is a finite, identifiable, contiguous, sequence of octets (e.g. a file or the content of an HTTP message or MIME attachment)
    * An indefinite-length, possibly non-finite sequence of octets is a stream
* "resource" is anything identified by an [absolute-URI](https://www.rfc-editor.org/rfc/rfc3986.html#section-4.3)
   * techincally this is a "primary resource", with a [resource identified by a fragment being a "secondary resource"](https://www.rfc-editor.org/rfc/rfc3986.html#section-3.5)
   * In JSON Schema, [any object containing an `$id` is a resource](https://www.ietf.org/archive/id/draft-bhutton-json-schema-01.html#section-4.3.5), as is any document whether it has an `$id` in the root object or not
 
Under these definitions, the differences between a resource and a document are that a document need not be identifiable _by a URI_, and a resource need not be finite (or even have a concrete representation at all, although OpenAPI is concerned with concrete resource representatoins).

Note that a document (which may or may not be a resource) can contain multiple resources within it.  `multipart` documents are the most obvious example, but JSON Schema documents where at least one subschema contains a `$id` are another, which means that OpenAPI 3.1 documents also can contain multiple resources.

A Schema Object that is the child of a different type of OAS Object and _does not_ contain a `$id` in that immediate Schema Object (it may or may not contain one in one or more subschema objects) is still part of the OpenAPI document _and resource_ in which it appears.  This is why you can use a JSON Pointer fragment with it, including into its subschemas (subject to [the caveats on JSON Pointer fragment crossing `$id` boundaries](https://www.ietf.org/archive/id/draft-bhutton-json-schema-01.html#name-json-pointer-fragments-and-) in the JSON Schema spec)

The JSON Schema spec uses "embedded" to talk about resources inside another resource; we should probably prefer RFC3986's "encapsulated" language instead.

**Reply by handrews** [reply]
A "description" can be one or many documents or resources.  An OAS 3.1 description can be one document while being many resources.

The way OAS 2.0 descriptions are specified, as interpreted to me by several TSC members (but disputed by @darrelmiller even for 2.0 IIRC) attempted to treat the "description" as a single resource split across multiple documents, _but did not do so in a consistent or coherent manner_.  For example, there is nothing anywhere in any spec that would allow JSON Pointer fragments to cross documents.  I suppose that could have been defined somehow, but it wasn't done and having a JSON Pointer cross what is self-apparently a document boundary (by the sequence of octets definition) is definitely odd and surprising.  idk if any 2.0 tools support it, I would find it baffling, but I find much about the 2.0 approach to references and documents and media types baffling.

**Reply by mikekistler** [reply]
Thanks @handrews  for your thoughts on this.

Regarding "resource", I think we should adopt the definition from [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-1.1): 
> the term "resource" is used in a general sense for whatever might be identified by a URI 

Note that RFC 3986 does not require the URI to be absolute, so fragments are allowed, but does not allow URI references -- that avoids the whole "Base URI" messiness. I want to note that RFC 3986 is consistent on this. In the [description of fragment](https://datatracker.ietf.org/doc/html/rfc3986#section-3.5) it says:

> The fragment identifier component of a URI allows indirect identification of a secondary resource by reference to a primary resource and additional identifying information.

If the thing identified by the fragment is a "secondary resource", then it is a "resource".

For the purposes of the OAS, I think we can constrain "document" to mean a "JSON or YAML document", since we explicitly state that an OAD:

> is itself a JSON object, which may be represented either in JSON or YAML format.

I know @handrews  feels that the definition of "document" in the [JSON Schema spec](https://json-schema.org/draft/2020-12/draft-bhutton-json-schema-01#section-4.1) comes from the wrong context, but I have trouble seeing any problems with it: 
> A JSON document is an information resource (series of octets) described by the application/json media type.
> In JSON Schema, the terms "JSON document", "JSON text", and "JSON value" are interchangeable because of the data model it defines.

And now "OpenAPI Description". The current 3.1.1 (dev) spec says:  

> An OpenAPI Description (OAD) formally describes the surface of an API and its semantics. It is composed of an entry document and any/all of its referenced documents. An OAD uses and conforms to the OpenAPI Specification, and MUST contain at least one paths field, components field, or webhooks field.

 This wording seems fine to me, but I do find it curious that it says "referenced documents" -- which (to me anyway) seems to imply that only documents that can be "referenced" are relevant, which by the definitions above implies that all relevant documents are also actually resources.

On the topic of fragments, @Henry raised the point that RFC 6901 clearly states:  

> In particular, the fragment identifier syntax for application/json is not JSON Pointer.

This was brand new to me! But it seems that the OAS actually overrides this (or at least I think a reader could reasonably conclude that it overrides it) with this statement in the [Relative References in URIs section](https://github.com/OAI/OpenAPI-Specification/blob/3.1.0/versions/3.1.0.md#relative-references-in-uris):

> If a URI contains a fragment identifier, then the fragment should be resolved per the fragment resolution mechanism of the referenced document. If the representation of the referenced document is JSON or YAML, then the fragment identifier SHOULD be interpreted as a JSON-Pointer as per RFC6901.

A further note on fragment interpretation is that according to [RFC 9512](https://datatracker.ietf.org/doc/html/rfc9512#section-1.2) the fragment identifier syntax for application/yaml _is_ JSON Pointer (along with YAML aliases). So we have to allow fragments to identify "referenced documents" within YAML files even without the override statement in the OAS.

And here let me pull one more quote from [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3.5):

> Fragment identifiers have a special role in information retrieval systems as the primary form of client-side indirect referencing, allowing an author to specifically identify aspects of an existing resource that are only indirectly provided by the resource owner.

So where does this leave us? 

I think the question that sparked this discussion was: when processing a reference _not in a Schema Object_ (we excluded that just to simply the discussion) that contains a fragment, can the document/resource/content at that reference be processed on its own, or must it be interpreted within the context of the larger primary resource in which it is contained?

Based on the content above (which folks are free to challenge), I have to conclude that the only requirement for processing the primary resource is for resolving the fragment to obtain the referenced resource, and that resource (and only that) is the "referenced document" that becomes part of the OpenAPI Description.

Now I also have to say that I'm not sure it's important to answer this question. Or said differently, I'm not sure what difference it makes one way or the other. If the reference were within a Schema Object, there could be implications since there may be $id fields in the primary resource that affect URI resolution within the secondary resource. But I'm not sure that is the case (until $self comes in 3.2) with references not in a Schema Object.

So maybe we need to consider if this is the right question to spend time debating?

**Reply by handrews** [reply]
@mikekistler yes, agreed, primary and secondary resources are both resources, thanks for that correction.

My definition and the JSON Schema definition are both in terms of "series of octets" and, _as far as the first sentence in the JSON Schema spec definition is concerns_, are consistent with each other.  The _second sentence_ is making a point about terms in the JSON RFC and is not relevant to the OAS.

> which by the definitions above implies that all relevant documents are also actually resources.

I think I agree.  Not 100% sure I'm not missing something, but I think I agree here.

> can the document/resource/content at that reference be processed on its own, or must it be interpreted within the context of the larger primary resource in which it is contained?

You can only process a secondary resource by determining the media type of the primary resource and applying that media type's fragment rules.  The fact that the OAS (and before that, JSON Reference) attempt to override that is irrelevant.  RFC3986 and other approved RFCs take precedence over random things someone attempted in a decade-plus-expired draft that was never reviewed properly.

**Reply by darrelmiller** [reply]
> You can only process a secondary resource by determining the media type of the primary resource and applying that media type's fragment rules. 

Where does that leave us when a media type, e.g. application/json does not provide any description of how to process fragment identifiers?  I think it suggests that using $ref to point to fragments of external documents is not valid.  That doesn't seem like a very tenable outcome.

**Reply by handrews** [reply]
@darrelmiller I didn't make the rules, although I don't see the problem.  You need to reference an `application/oas+json` or `application/schema+json` document to use the fragment syntax (and if you don't have a `Content-Type` header you can infer the media type from other cues including extensions and content-sniffing).  For JSON Hyper-Schema, in order to use fragments with instance documents, we invented `application/schema-instance+json`.

**Reply by mikekistler** [reply]
I'm not sure any rules are being broken by interpreting a fragment as a JSON pointer for JSON documents.

RFC 6901 just says

> In particular, the fragment identifier syntax for application/json is not JSON Pointer.

This is a true statement -- there is _no_ defined fragment identifier syntax for application/json in the [IANA Media Types registry entry for application/json](https://www.iana.org/assignments/media-types/application/json).

But I found no language in RFC 3986 that _prohibits_ applying a fragment identifier syntax in the event that the media type does not define one. I did find this statement:

> The fragment's format and resolution is therefore dependent on the media type [RFC2046] of a potentially retrieved representation, even though such a retrieval is only performed if the URI is dereferenced.  If no such representation exists, then the semantics of the fragment are considered unknown and are effectively unconstrained.

I'm having a little trouble parsing this but in any event it is clear that in _some_ situations RFC 3986 allows the fragment to be interpreted in an "unconstrained" way.

tl;dr: The OAS says "here's how you handle this otherwise undefined situation" -- I don't see that as breaking a rule.

**Reply by handrews** [reply]
@mikekistler We are defining media types for OAS and JSON Schema.  They are not properly registered, but conceptually it all works.  We don't need to come up with novel ways of handling this.

The only thing that _definitely_ goes against things is _disregarding_ a media type, including `application/json`.  Although I consider it perfectly valid to _refine_ `application/json` to `application/oas+json` (or whatever it is supposed to be) by content-sniffing.  But it's the "I'm not even going to acknowledge that this primary resource _has_ a media type so I can do weird non-standard things with extracting fragments" that is a problem.  That's not a case where things are ambiguous and you're filling in details.  That's willfully ignoring the information that is there.

**Comment by darrelmiller** [comment]
@mikekistler 
> which (to me anyway) seems to imply that only documents that can be "referenced" are relevant,

I'm not sure you should correlate "referenced" with the use of "$ref".  There are other ways to reference documents from an OpenAPI description.  My interpretation is that a $id that identifies a schema in another document effectively "references" that document.  

> So we have to allow fragments to identify "referenced documents" within YAML files

This statement makes me very squirrely.  I don't like the idea of using the word "document" for fragments of a "physical" document.

> I have to conclude that the only requirement for processing the primary resource is for resolving the fragment to obtain the referenced resource, and that resource (and only that) is the "referenced document" that becomes part of the OpenAPI Description.

Is the purpose of this statement to try and clarify that the server of a referenced operation comes from the entry document and not the host document? (and other things like tags and security).

**Comment by mikekistler** [comment]
@darrelmiller 

> > So we have to allow fragments to identify "referenced documents" within YAML files

> This statement makes me very squirrely. I don't like the idea of using the word "document" for fragments of a "physical" document.

I think this is part of the terminology we need to work out. It seems that you would prefer to use the term "fragment" for "secondary resources within a primary resource" (as RFC 3986 would put it). I suppose we _could_ do that. That might require a scrub of the spec to determine where current uses of the term "document" should instead be "fragment", though a quick scan suggests there wouldn't be many of these.

But a "fragment" would still be logically a JSON object, right? Which by the JSON Schema terminology is a "JSON Document". So we start to chase our tails a bit if we want to distinguish between "document" and "fragment".

And isn't it possible, for schema objects anyway, to convert any "fragment" into a "document" by adding a "$id" with an absolute URI to it? So again, the same thing can be either a document or fragment, depending only on how it is referenced (I think).

> Is the purpose of this statement to try and clarify that the server of a referenced operation comes from the entry document and not the host document? (and other things like tags and security).

Yes but even more relevant is that if there happens to be an `openapi` element in the "primary resource", it should have no bearing on how a "secondary resource" -- "fragment" in your terminology -- is processed.

**Reply by handrews** [reply]
@mikekistler I'm losing track of what you're trying to do.  I keep trying to offer an RFC-based, standards-compliant, coherent view of the world and it feels like you are looking for some sort of alternative.

Why?  Why not just go with what fits with everything else?  It's one thing if there's a point I'm making that is unclear, but it is starting to feel to me like you're hunting for places where you can find alternate interpretations.  Which maybe I could help with more if I understood your goal.  Right now I feel like you're tossing out a lot of speculative possibilities and I can't figure out how to respond because I don't know why.

Particularly because this is the discussion about 3.1.  I feel very strongly that 3.1 works as-is.  The only confusions come from 3.0 (and 2.0).  There's nothing about URIs or media types or fragments that need fixing in 3.1.  The whole "extract a JSON object out of an unknown document" thing will never work in 3.1.  We already agreed it wouldn't.  There is cleanup to do around "document" but there's no need for a new conceptual framework in 3.1.

So what is it that you're trying to find with all of this?

**Reply by handrews** [reply]
> And isn't it possible, for schema objects anyway, to convert any "fragment" into a "document" by adding a "$id" with an absolute URI to it? So again, the same thing can be either a document or fragment, depending only on how it is referenced (I think).

This confuses document and resource.  An `$id` in a subschema makes that subschema into a resource.  A document is a coherent, finite sequence of octets, like a file.  You don't point to a subset of a file and also call that a file no matter what keywords it is using.  A file is a chunk of storage of some sort.

**Comment by handrews** [comment]
@mikekistler @darrelmiller did we sort this out sufficiently?  Do we need to keep this open?  (closed discussions, like closed issues, are still easily find-able).

**Comment by handrews** [comment]
Given all the work we did here and the lack of response to my "can we close this?" I'm going to close it :-)
